server/src/main/java/org/elasticsearch/script/field/vectors/ByteBinaryDenseVector.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.script.field.vectors;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.core.SuppressForbidden;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+public class ByteBinaryDenseVector implements DenseVector {
+
+    public static final int MAGNITUDE_BYTES = 4;
+
+    protected final BytesRef docVector;
+    protected final int dims;
+
+    protected float[] floatDocVector;
+    protected byte[] byteDocVector;
+    protected boolean magnitudeDecoded;
+    protected float magnitude;
+
+    public ByteBinaryDenseVector(BytesRef docVector, int dims) {
+        this.docVector = docVector;
+        this.dims = dims;
+    }
+
+    @Override
+    public float[] getVector() {
+        if (floatDocVector == null) {
+            floatDocVector = new float[dims];
+
+            int i = 0;
+            int j = docVector.offset;
+            while (i < dims) {
+                floatDocVector[i++] = docVector.bytes[j++];
+            }
+        }
+
+        return floatDocVector;
+    }
+
+    @Override
+    public byte[] asBytes() {
+        if (byteDocVector == null) {
+            byteDocVector = new byte[dims];
+            System.arraycopy(docVector.bytes, docVector.offset, byteDocVector, 0, dims);
+        }
+
+        return byteDocVector;
+    }
+
+    @Override
+    public float getMagnitude() {
+        if (magnitudeDecoded == false) {
+            magnitude = ByteBuffer.wrap(docVector.bytes, docVector.offset + dims, MAGNITUDE_BYTES).getFloat();
+            magnitudeDecoded = true;
+        }
+        return magnitude;
+    }
+
+    @Override
+    public double dotProduct(byte[] queryVector) {
+        int result = 0;
+        int i = 0;
+        int j = docVector.offset;
+        while (i < dims) {
+            result += docVector.bytes[j++] * queryVector[i++];
+        }
+        return result;
+    }
+
+    @Override
+    public double dotProduct(float[] queryVector) {
+        int result = 0;
+        int i = 0;
+        int j = docVector.offset;
+        while (i < dims) {
+            result += docVector.bytes[j++] * (int) queryVector[i++];
+        }
+        return result;
+    }
+
+    protected double dotProductNormalized(float[] normalizedQueryVector) {
+        float result = 0;
+        int i = 0;
+        int j = docVector.offset;
+        while (i < dims) {
+            result += docVector.bytes[j++] * normalizedQueryVector[i++];
+        }
+        return result;
```

Comment:
Agreed on the trickiness of this. I would like to add this to a high-bandwidth discussion topic as well. I wasn't really sure what to do here exactly, but it really did not seem valuable to have a normalized `byte` vector.

Reply:
Another discussion item around this as a similar topic -- it makes sense to check the query vector for correct bounds in class bindings because it's only called once per segment and cached. But if a user calls one of the methods directly on a `DenseVector` as a `field('my-vector').get().dotProduct(new float[] {...})` then checking could be a large performance hit. Right now we just cast, so the user is responsible for giving a valid vector in this particular case.

Reply:
Updated do stop mixed operations between byte[] and float[] vectors.